timemachine/cpp/src/gpu_utils.cuh

```diff
@@ -83,3 +84,85 @@ double __device__ __forceinline__ rmul_rn(double a, double b) { return __dmul_rn
 float __device__ __forceinline__ radd_rn(float a, float b) { return __fadd_rn(a, b); }
 
 double __device__ __forceinline__ radd_rn(double a, double b) { return __dadd_rn(a, b); }
+
+// If this were not int128s, we could do __shfl_down_sync, but only supports up to 64 values
+__device__ __forceinline__ void energy_warp_reduce(volatile __int128 *shared_data, int thread_idx) {
+#pragma unroll 6
+    for (int i = 32; i > 0; i /= 2) {
+        if (thread_idx < i) {
+            shared_data[thread_idx] = shared_data[thread_idx] + shared_data[thread_idx + i];
+        }
+        __syncwarp();
+    }
+}
+
+template <unsigned int THREADS_PER_BLOCK>
+__device__ __forceinline__ void block_energy_reduce(volatile __int128 *shared_data, int tid) {
+    static_assert(THREADS_PER_BLOCK <= 1024 && (THREADS_PER_BLOCK & (THREADS_PER_BLOCK - 1)) == 0);
+
+    // For larger blocks larger than a warp
+    if (THREADS_PER_BLOCK >= 1024) {
+        if (tid < 512) {
+            shared_data[tid] += shared_data[tid + 512];
+        }
+        __syncthreads();
+    }
+    if (THREADS_PER_BLOCK >= 512) {
+        if (tid < 256) {
+            shared_data[tid] += shared_data[tid + 256];
+        }
+        __syncthreads();
+    }
+    if (THREADS_PER_BLOCK >= 256) {
+        if (tid < 128) {
+            shared_data[tid] += shared_data[tid + 128];
+        }
+        __syncthreads();
+    }
+    if (THREADS_PER_BLOCK >= 128) {
+        if (tid < 64) {
+            shared_data[tid] += shared_data[tid + 64];
+        }
+        __syncthreads();
+    }
+
+    if (tid < WARP_SIZE) {
+        energy_warp_reduce(shared_data, tid);
+    }
+}
+
+// Taken originally from https://developer.download.nvidia.com/assets/cuda/files/reduction.pdf
+// but modified to account for post Cuda 9.0 which no longer implicitly syncs warps and the
+// fact that we don't need to launch recursive blocks as we can simply reduce the number of energies
+// that are accumulated.
+template <unsigned int BLOCK_SIZE>
+void __global__ k_accumulate_energy(
+    int N,
+    const __int128 *__restrict__ input_buffer, // [N]
+    __int128 *__restrict__ u_buffer            // [1]
+) {
+
+    __shared__ __int128 shared_mem[BLOCK_SIZE];
+    unsigned int tid = threadIdx.x;
+    if (tid >= BLOCK_SIZE) {
+        return;
+    }
+
+    unsigned int i = tid;
+    const unsigned int stride = BLOCK_SIZE * 2;
+    shared_mem[tid] = 0;
+    while (i < N) {
+        // Only add two values if the input buffer has enough input
+        shared_mem[tid] += input_buffer[i] + (i + BLOCK_SIZE < N ? input_buffer[i + BLOCK_SIZE] : 0);
+        i += stride;
+    }
+    //
+    __syncthreads();
+
+    block_energy_reduce<BLOCK_SIZE>(shared_mem, threadIdx.x);
+    if (tid == 0) {
+        // This could be a race condition if multiple `k_accumulate_energy` are running
+        // on the same u_buffer
+        u_buffer[0] = shared_mem[0];
```

Comment:
Is this comment actionable?

Reply:
I believe so? You could easily write the code in a way that would result in different potentials clobbering the energy buffers

Reply:
>You could easily write the code in a way that would result in different potentials clobbering the energy buffers

Sorry, I don't have a good sense of context or severity.

(Can this already happen (even if rarely or non-deterministically) for any of the potential configurations in the tests? Or is this just a possible future problem, which could only occur if, say, the implementation of `StreamedPotentialRunner` is modified unsafely? If a current problem, would atomically incrementing `u_buffer[0]` avoid it?)